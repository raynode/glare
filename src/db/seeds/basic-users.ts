import { NodeType } from 'gram'
import * as Knex from 'knex'
import { User } from '../models'

export const seed = async (knex: Knex) => {
  Promise.all(['Users', 'Levels', 'Worlds', 'Games'].map(table => knex(table).del()))

  type RequiredProperties = 'name' | 'email' | 'state'
  const users: Array<Pick<User, RequiredProperties> & Partial<User>> = [
    {
      // ADMIN
      givenName: 'Kopelke',
      nickname: 'Nox',
      name: 'nox',
      state: 'admin',
      email: 'kopelke@gmail.com',
      emailVerified: true,
    },
    {
      // MEMBER
      nickname: 'Testy',
      name: 'Test-User',
      state: 'member',
      email: 'nox@raynode.de',
      emailVerified: true,
    },
    {
      // GUEST
      name: 'unknown',
      state: 'guest',
      email: 'unknown@example.com',
      emailVerified: false,
    },
    {
      name: 'RETRACTED',
      state: 'guest',
      email: 'retracted@raynode.de',
      emailVerified: true,
      deleted_at: new Date(),
    },
  ]

  await knex('Users').insert(users)
  await knex('Games').insert({
    name: 'FindPuzzle',
  })
  const findPuzzle = (await knex('Games').where({ name: 'FindPuzzle' }))[0]

  await knex('Worlds').insert({
    name: 'Sehr Einfach',
    gameId: findPuzzle.id,
  })

  const easyPuzzles = (await knex('Worlds').where({ name: 'Sehr Einfach' }))[0]
  await knex('Levels').insert({
    name: '1 - Level',
    gameId: findPuzzle.id,
    worldId: easyPuzzles.id,
    data: {
      actions: ['cellToEmpty', 'cellToUnknown', 'cellToRabbit'],
      cols: 6,
      rows: 6,
      index: [
        'ðŸ¥•',
        '?',
        '?',
        '?',
        'ðŸ¥•',
        '?',
        '?',
        '?',
        '?',
        'ðŸ¥•',
        '?',
        '?',
        '?',
        '?',
        '?',
        'ðŸ¥•',
        '?',
        '?',
        '?',
        '?',
        '?',
        '?',
        '?',
        '?',
        '?',
        'ðŸ¥•',
        '?',
        '?',
        'ðŸ¥•',
        '?',
        '?',
        '?',
        'ðŸ¥•',
        '?',
        '?',
        '?',
      ],
      rowHints: [3, 0, 1, 1, 0, 2],
      colHints: [0, 3, 0, 1, 2, 1],
    },
  })
}
