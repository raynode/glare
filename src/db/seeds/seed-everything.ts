import { NodeType } from 'gram'
import * as Knex from 'knex'
import { User } from '../models'

import { findPuzzle } from 'nx-gameboard'

// convert the symbols into numbers
export const [empty, target, hint, unknown] = [' ', '🐇', '🥕', '?']

export const createFindPuzzleLevel = (seed: string, size: number) => {
  const findPuzzleBoard = findPuzzle.RabbitChaseBoard({
    seed,
    cols: size,
    rows: size,
  })
  findPuzzleBoard.init()
  const solution = findPuzzleBoard.getInitialSolution()
  return {
    ...solution,
    index: solution.index.map(
      val =>
        ({
          [empty]: 0,
          [target]: 1,
          [hint]: 2,
          [unknown]: 3,
        }[val]),
    ),
  }
}

export const seed = async (knex: Knex) => {
  Promise.all(['Users', 'Levels', 'Worlds', 'Games'].map(table => knex(table).del()))

  type RequiredProperties = 'name' | 'email' | 'state'
  const users: Array<Pick<User, RequiredProperties> & Partial<User>> = [
    {
      // ADMIN
      givenName: 'Kopelke',
      nickname: 'Nox',
      name: 'nox',
      state: 'admin',
      email: 'kopelke@gmail.com',
      emailVerified: true,
    },
    {
      // MEMBER
      nickname: 'Testy',
      name: 'Test-User',
      state: 'member',
      email: 'nox@raynode.de',
      emailVerified: true,
    },
    {
      // GUEST
      name: 'unknown',
      state: 'guest',
      email: 'unknown@example.com',
      emailVerified: false,
    },
    {
      name: 'RETRACTED',
      state: 'guest',
      email: 'retracted@raynode.de',
      emailVerified: true,
      deleted_at: new Date(),
    },
  ]

  await knex('Users').insert(users)
  await knex('Games').insert({
    name: 'FindPuzzle',
  })
  const findPuzzle = (await knex('Games').where({ name: 'FindPuzzle' }))[0]

  const createWorld = async (name: string) =>
    (await knex('Worlds')
      .insert({
        name,
        gameId: findPuzzle.id,
      })
      .returning('id'))[0]

  const createLevels = (worldId: string, size: number) =>
    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10].reduce(
      (promise, lvl) =>
        promise.then(() =>
          knex('Levels').insert({
            name: `${lvl} - Level`,
            gameId: findPuzzle.id,
            worldId,
            data: createFindPuzzleLevel(`${lvl} - Level`, size),
          }),
        ),
      Promise.resolve(),
    )

  const createWorldLevels = async ([name, size]: [string, number]) => createLevels(await createWorld(name), size)

  const createWorlds = () =>
    [
      ['Sehr leicht', 4],
      ['Leicht', 5],
      ['Standard', 6],
      ['Schwer', 7],
      ['Schwerer', 8],
      ['Noch Schwerer', 9],
      ['Wirklich?', 12],
    ].reduce(
      (promise: Promise<any>, worldDef: [string, number]) => promise.then(() => createWorldLevels(worldDef)),
      Promise.resolve(),
    )

  await createWorlds()
}
