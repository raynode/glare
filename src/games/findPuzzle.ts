import { CreateGameSolution } from 'db/models'

export enum CellStates {
  empty,
  target,
  hint,
  unknown,
}

export enum SolutionStates {
  unchanged,
  active,
  won,
  lost,
}

export interface Action<Types extends string = string, Payload = any> {
  type: Types
  payload: Payload
}

/** Checks if the action has a valid FindPuzzleAction type */
export const isValidFindPuzzleAction = (action: Action): action is FindPuzzleAction =>
  ['cellToEmpty', 'cellToUnknown', 'cellToRabbit'].includes(action.type)

/** Generator to create a isAction guard */
export const isActionType = <ActionType extends Action>(type: string) => (action: Action): action is ActionType =>
  action.type === type

export type FindPuzzleActionCellToEmpty = Action<'cellToEmpty', { index: number }>
export const isFindPuzzleActionCellToEmpty = isActionType<FindPuzzleActionCellToEmpty>('cellToEmpty')

export type FindPuzzleActionCellToUnknown = Action<'cellToUnknown', { index: number }>
export const isFindPuzzleActionCellToUnknown = isActionType<FindPuzzleActionCellToUnknown>('cellToUnknown')

export type FindPuzzleActionCellToTarget = Action<'cellToRabbit', { index: number }>
export const isFindPuzzleActionCellToTarget = isActionType<FindPuzzleActionCellToTarget>('cellToRabbit')

export type FindPuzzleAction =
  | FindPuzzleActionCellToEmpty
  | FindPuzzleActionCellToUnknown
  | FindPuzzleActionCellToTarget

export interface FindPuzzleSolutionData {
  index: CellStates[]
  actions: string[]
  cols: number
  rows: number
  rowHints: number[]
  colHints: number[]
}

export interface FindPuzzleHistory {
  action: FindPuzzleAction
}

export interface FindPuzzleSolution extends CreateGameSolution {
  history: FindPuzzleHistory[]
  data: FindPuzzleSolutionData
  state: SolutionStates
}

export const isValidIndex = (solution: FindPuzzleSolution, index: number) =>
  index >= 0 && index < solution.data.index.length

/** checks if a cellstate change is valid */
export const validateCellStateChange = (solution: FindPuzzleSolution, index: number, validCellStates: CellStates[]) => {
  if (!isValidIndex(solution, index)) return false
  const currentIndex = solution.data.index[index]
  if (!validCellStates.includes(currentIndex)) return false
  return true
}

/** changes a specific index to a different cellstate */
export const updateCellState = (
  solutionData: FindPuzzleSolutionData,
  index: number,
  cellState: CellStates,
): FindPuzzleSolutionData => ({
  ...solutionData,
  index: [...solutionData.index.slice(0, index), cellState, ...solutionData.index.slice(index + 1)],
})

/** Finds the number of target values in a array of CellStates */
export const indexToTargetNumber = (index: CellStates[]) => index.filter(state => state === CellStates.target).length

/** Generates a function that sums up all targets for a given row */
export const getIndexRow = (cols: number, rows: number, index: CellStates[]) => (row: number) =>
  indexToTargetNumber(index.filter((state, index) => row === index % rows))

/** Generates a function that sums up all targets for a given column */
export const getIndexCol = (cols: number, rows: number, index: CellStates[]) => (col: number) => {
  const [lower, upper] = [rows * col, rows * (col + 1)]
  return indexToTargetNumber(index.filter((state, index) => lower <= index && index < upper))
}

/**
 * Runs through a solution and determines the solutions state
 * - if not history, no actions were taken => unchanged
 * - if no field is unknown...
 *   - if all row-hints and col-hints are correct in their numbers => won
 *   - else => lost
 * - else => active
 */
export const resolveState = (solution: FindPuzzleSolution) => {
  if (!solution.history.length) return SolutionStates.unchanged

  if (!solution.data.index.filter(state => state === CellStates.unknown).length) {
    const indexRow = getIndexRow(solution.data.cols, solution.data.rows, solution.data.index)
    const indexCol = getIndexCol(solution.data.cols, solution.data.rows, solution.data.index)
    return solution.data.rowHints
      .map((hint, row) => hint === indexRow(row))
      .concat(solution.data.colHints.map((hint, col) => hint === indexCol(col)))
      .filter(correct => !correct).length
      ? SolutionStates.lost
      : SolutionStates.won
  }

  return SolutionStates.active
}

/** checks the solution for its current state */
export const check = (solution: FindPuzzleSolution): FindPuzzleSolution => {
  return {
    ...solution,
    state: resolveState(solution),
  }
}

/** resolves a solution for an change-board action */
export const reduceChangeAction = (
  solution: FindPuzzleSolution,
  action: FindPuzzleAction,
  validCellStates: CellStates[],
  newCellState: CellStates,
): FindPuzzleSolution =>
  !validateCellStateChange(solution, action.payload.index, validCellStates)
    ? solution
    : check({
        ...solution,
        data: updateCellState(solution.data, action.payload.index, newCellState),
        history: [...solution.history, { action }],
      })

/** resolves all FindPuzzleActions */
export const reducer = (solution: FindPuzzleSolution, action: FindPuzzleAction) => {
  if (!isValidFindPuzzleAction(action)) return solution // never
  if (isFindPuzzleActionCellToEmpty(action))
    return reduceChangeAction(solution, action, [CellStates.unknown, CellStates.target], CellStates.empty)
  if (isFindPuzzleActionCellToTarget(action))
    return reduceChangeAction(solution, action, [CellStates.unknown, CellStates.empty], CellStates.target)
  if (isFindPuzzleActionCellToUnknown(action))
    return reduceChangeAction(solution, action, [CellStates.empty, CellStates.target], CellStates.unknown)
  return solution // never
}
