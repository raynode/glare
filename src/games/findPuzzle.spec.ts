import { createFindPuzzleLevel } from 'db/seeds/seed-everything'
import { CellStates, FindPuzzleSolution, FindPuzzleSolutionData, reducer, SolutionStates } from './findPuzzle'

const createSolution = (data: FindPuzzleSolutionData): FindPuzzleSolution => ({
  history: [],
  levelId: 'dummy',
  playerId: 'dummy',
  state: SolutionStates.unchanged,
  data,
})

const createSolutionData = (size: number = 4): FindPuzzleSolutionData => createFindPuzzleLevel('TEST', size)

describe('Game:Find Puzzle', () => {
  let solution: FindPuzzleSolution = null

  beforeEach(() => (solution = createSolution(createSolutionData())))

  const expectChanges = (next: FindPuzzleSolution) => {
    expect(solution.history).not.toEqual(next.history)
    expect(solution.data.actions).toEqual(next.data.actions)
    expect(solution.data.colHints).toEqual(next.data.colHints)
    expect(solution.data.rowHints).toEqual(next.data.rowHints)
    expect(solution.data.cols).toEqual(next.data.cols)
    expect(solution.data.rows).toEqual(next.data.rows)
    expect(solution.data.index).not.toEqual(next.data.index)
  }

  it('should not change the level!', () => expect(solution).toMatchSnapshot())
  it('should generate lvl 3', () => expect(createSolution(createSolutionData(3))).toMatchSnapshot())
  it('should generate lvl 2', () => expect(createSolution(createSolutionData(2))).toMatchSnapshot())

  it('', () => console.log(solution))

  it('should handle cellToEmpty actions correctly', () => {
    const next = reducer(solution, {
      type: 'cellToEmpty',
      payload: {
        index: 0,
      },
    })
    expectChanges(next)
    expect(solution.data.index.slice(1)).toEqual(next.data.index.slice(1))
    expect(next.data.index[0]).toEqual(CellStates.empty)
    expect(next.history).toHaveLength(1)
  })

  it('should handle cellToUnkown actions correctly', () => {
    const next = reducer(solution, {
      type: 'cellToEmpty',
      payload: {
        index: 0,
      },
    })
    const nextNext = reducer(next, {
      type: 'cellToUnknown',
      payload: {
        index: 0,
      },
    })
    expectChanges(next)
    expect(solution.data).toEqual(nextNext.data)
    expect(nextNext.history).toHaveLength(2)
    expect(nextNext.data.index[0]).toEqual(CellStates.unknown)
  })

  it('should handle cellToEmpty actions correctly', () => {
    const next = reducer(solution, {
      type: 'cellToRabbit',
      payload: {
        index: 0,
      },
    })
    expectChanges(next)
    expect(next.history).toHaveLength(1)
    expect(solution.data.index.slice(1)).toEqual(next.data.index.slice(1))
    expect(next.data.index[0]).toEqual(CellStates.target)
  })
})

describe('Winning and Loosing', () => {
  let solution: FindPuzzleSolution = null
  beforeEach(() => (solution = createSolution(createSolutionData(2))))

  it('should solve an easy puzzle', () => {
    const step1 = reducer(solution, {
      type: 'cellToEmpty',
      payload: { index: 0 },
    })
    const step2 = reducer(step1, {
      type: 'cellToEmpty',
      payload: { index: 2 },
    })
    const step3 = reducer(step2, {
      type: 'cellToRabbit',
      payload: { index: 3 },
    })
    // unchanged, active, active, won
    expect([solution, step1, step2, step3].map(s => s.state)).toEqual([0, 1, 1, 2])
  })
})
