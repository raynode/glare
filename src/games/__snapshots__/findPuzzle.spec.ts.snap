// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game:Find Puzzle should generate lvl 2 1`] = `
Object {
  "data": Object {
    "actions": Array [
      "cellToEmpty",
      "cellToUnknown",
      "cellToRabbit",
    ],
    "colHints": Array [
      0,
      1,
    ],
    "cols": 2,
    "index": Array [
      3,
      2,
      3,
      3,
    ],
    "rowHints": Array [
      0,
      1,
    ],
    "rows": 2,
  },
  "history": Array [],
  "levelId": "dummy",
  "state": 0,
  "userId": "dummy",
}
`;

exports[`Game:Find Puzzle should generate lvl 3 1`] = `
Object {
  "data": Object {
    "actions": Array [
      "cellToEmpty",
      "cellToUnknown",
      "cellToRabbit",
    ],
    "colHints": Array [
      2,
      0,
      2,
    ],
    "cols": 3,
    "index": Array [
      3,
      2,
      3,
      2,
      3,
      2,
      3,
      2,
      3,
    ],
    "rowHints": Array [
      2,
      0,
      2,
    ],
    "rows": 3,
  },
  "history": Array [],
  "levelId": "dummy",
  "state": 0,
  "userId": "dummy",
}
`;

exports[`Game:Find Puzzle should not change the level! 1`] = `
Object {
  "data": Object {
    "actions": Array [
      "cellToEmpty",
      "cellToUnknown",
      "cellToRabbit",
    ],
    "colHints": Array [
      1,
      1,
      1,
      1,
    ],
    "cols": 4,
    "index": Array [
      3,
      2,
      3,
      3,
      3,
      3,
      2,
      3,
      3,
      2,
      3,
      3,
      3,
      3,
      2,
      3,
    ],
    "rowHints": Array [
      2,
      0,
      0,
      2,
    ],
    "rows": 4,
  },
  "history": Array [],
  "levelId": "dummy",
  "state": 0,
  "userId": "dummy",
}
`;
